File: c++/704. Binary Search/704.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if(nums.size() == 0) //if list empty
            return -1;
        int l = 0; //first element index
        int r = nums.size(); //last element + 1 index
        while (l+1<r){ //while this
            int m = l + (r-l)/2; //we search mid
            if(nums[m]>target){ //if nums[mid] > target
                r=m; //now our mid is right element 
            }else{
                l=m; //else our mid is left element
            }
        }return nums[l]==target?l:-1; //and if this single elemnt == target we return this index
    }
};

File: python/704. Binary Search/704.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 0: #if its empty we dont even need to to check something
            return -1
        l = 0 #lets define left like our first element index
        r = len(nums) #and lets define right our last elemnt + 1 index
        while l+1<r: #so we searching mid when its at least 3 elements in list 
            m = l+(r-l)//2 #search mid
            if nums[m]>target: #if elemnt with m index > target 
                r=m #we should update right
            else: 
                l=m #else - left
        if nums[l] == target: #so if our nums[l] (updated l) we gonna equal with our target
            return l #and if its ok we gonna return this
        else:
            return -1
            
